src/attr/_next_gen.py

```diff
@@ -73,11 +74,37 @@ def do_it(cls, auto_attribs):
             on_setattr=on_setattr,
         )
 
-    if auto_attribs is not None:
-        return do_it(maybe_cls, auto_attribs)
-
     def wrap(cls):
-        # Making this a wrapper ensures this code runs during class creation.
+        """
+        Making this a wrapper ensures this code runs during class creation.
+
+        We also ensure that frozen-ness of classes is inherited.
+        """
+        nonlocal frozen, on_setattr
+
+        had_on_setattr = on_setattr is not None
+
+        # By default, mutable classes validate on setattr.
+        if frozen is False and on_setattr is None:
+            on_setattr = setters.validate
+
+        # However, if we subclass a frozen class, we inherit the immutability
+        # and disable on_setattr.
+        for base_cls in cls.__bases__:
+            if base_cls.__setattr__ is _frozen_setattrs:
+                if had_on_setattr:
+                    raise ValueError(
+                        "Frozen classes can't use on_setattr "
+                        "(frozen-ness was inherited)."
+                    )
+
+                frozen = True
```

Comment:
Just 2 small questions.  
Couldn't this be just
```
on_setattr = None
```

I believe frozen will be inherited anyway right?  Or are you figuring well we've already spent time detecting it so we might as well use it?

Reply:
You're right, let's keep it auto-detected.